programs/perpetuals/tests/anchor/test_client.ts

```diff
@@ -706,6 +708,88 @@ export class TestClient {
     }
   };
 
+  setCustomOraclePricePermissionless = async (
+    permissionlessOracleKeypair: Keypair,
+    price: number,
+    custody,
+    publishTime?,
+    noSignatureVerification?,
+    messageOverwrite?
+  ) => {
+    const schema = {
+      struct: {
+        custodyAccount: { array: { type: "u8", len: 32 } },
+        price: "u64",
+        expo: "i32",
+        conf: "u64",
+        ema: "u64",
+        publishTime: "i64",
+      },
+    };
+
+    let setCustomOraclePricePermissionlessParams = {
+      custodyAccount: custody.custody,
+      price: new BN(price * 1000),
+      expo: -3,
+      conf: new BN(0),
+      ema: new BN(price * 1000),
+      publishTime:
+        publishTime != null ? new BN(publishTime) : new BN(this.getTime()),
+    };
+
+    let message =
+      messageOverwrite != null
+        ? messageOverwrite
+        : borsh.serialize(schema, {
+            ...setCustomOraclePricePermissionlessParams,
+            custodyAccount:
+              setCustomOraclePricePermissionlessParams.custodyAccount.toBytes(),
+          });
+
+    const signature = nacl.sign.detached(
+      message,
+      permissionlessOracleKeypair.secretKey
+    );
+    let tx = new anchor.web3.Transaction();
+    if (noSignatureVerification == null) {
+      tx.add(
+        anchor.web3.Ed25519Program.createInstructionWithPublicKey({
+          publicKey: permissionlessOracleKeypair.publicKey.toBytes(),
+          message: message,
+          signature: signature,
+        })
+      );
+    }
+    tx.add(
+      this.program.instruction.setCustomOraclePricePermissionless(
+        setCustomOraclePricePermissionlessParams,
+        {
+          accounts: {
+            perpetuals: this.perpetuals.publicKey,
+            pool: this.pool.publicKey,
+            custody: custody.custody,
+            oracleAccount: custody.oracleAccount,
+            systemProgram: SystemProgram.programId,
+            ixSysvar: anchor.web3.SYSVAR_INSTRUCTIONS_PUBKEY,
+          },
+        }
+      )
+    );
+
+    try {
```

Comment:
instead of building a custom instruction you could use `.preInstructions` , i.e. `this.program.methods.xxx(...).accounts(...)..preInstructions(your_ed2519_ix).rpc()`

Reply:
yes done!  built tx with `.preInstructions`

Reply:
could you confirm that `preInstructions()` call adds instructions to the tx and not replaces the ones that were stored there before? Otherwise you would replace the call to Ed25519 program with set compute budget

Reply:
Yes, added a print statement and saw that the successive `tx = tx.preInstructions(...)` calls add to each other
<img width="688" alt="Screenshot 2023-10-14 at 5 27 19 PM" src="https://github.com/solana-labs/perpetuals/assets/1387955/9b8b85c7-911a-40a6-a127-245270b41396">

Also have a test case where the ed25519 signature is missing -- confirms that the tx fails with `PermissionlessOracleMissingSignature`.